### PR TITLE
Overflow protection and rounding to +- inf for Add function

### PR DESCRIPTION
**Pull request by Joshcarp**
as mentioned in issue #11

### CODE REVIEW

File: decimal64_test.go [hunk 1]
```diff
@@ -1,10 +1,10 @@
 package decimal
 
 import (
+	//"fmt"
```
Comment (anzdaddy) [inline decimal64_test.go]: Remove the comment. If you're keeping it around for convenience, a properly setup editor will add these automatically so you don't need to.

File: decimal64_test.go [hunk 2]
```diff
@@ -1,10 +1,10 @@
...
+	"github.com/stretchr/testify/require"
```
Comment (anzdaddy) [inline decimal64_test.go]: Original placement was better.

File: decimal64math_test.go [hunk 3]
```diff
@@ -370,3 +370,18 @@ func BenchmarkDecimal64Sub(b *testing.B) {
 		_ = x[i%len(x)].Sub(y[i%len(y)])
 	}
 }
+
+
+func TestBoundaries(t *testing.T) {
+	require := require.New(t)
+	x := MustParseDecimal64("9e384")
+	fmt.Println("x.parts()")
+	fmt.Println(x.parts())
```
Comment (anzdaddy) [inline decimal64math_test.go]: Don't print things out in tests. Check their values.

File: decimal64math.go [hunk 4]
```diff
@@ -25,11 +25,18 @@ func (d Decimal64) Add(e Decimal64) Decimal64 {
...
+
 	if sign1 == sign2 {
 		significand := significand1 + significand2
-		if significand >= decimal64Base {
+
+		// avoid overflow and round to inf
+		if significand >= decimal64Base && exp1 < expMax {
 			exp1++
 			significand /= 10
+
```
Comment (anzdaddy) [inline decimal64math.go]: Unnecessary blank line

File: decimal64math.go [hunk 5]
```diff
@@ -25,11 +25,18 @@ func (d Decimal64) Add(e Decimal64) Decimal64 {
...
+		} else if significand >= decimal64Base {
+
```
Comment (anzdaddy) [inline decimal64math.go]: Unnecessary blank line

File: decimal64math_test.go [hunk 6]
```diff
@@ -1,9 +1,8 @@
 package decimal
 
 import (
-	"testing"
-
 	"github.com/stretchr/testify/require"
+	"testing"
```
Comment (anzdaddy) [inline decimal64math_test.go]: Original placement was better.

**Comment by juliaogris** [comment]
Heyhey, first approval by @anzdaddy - took me a lot longer to get there.
Good work.
One minor things re commit messages, please [have a read](https://chris.beams.io/posts/git-commit/), especially:
> Capitalize the subject line
Use the imperative mood in the subject line

ie `Add tests` rather than `added tests` - I can imagine that sounds super pedantic to you but it really makes a difference when you go through the git history to try and work out where a bug was introduced....

File: decimal64_test.go (hunk 2 above)
Comment (juliaogris) [reply decimal64_test.go]: `goimports` and `gofmt` should ideally be executed on every save. You can probably set your editor up this way so you don't have to worry about it.

File: decimal64math.go (hunk 4 above)
Comment (juliaogris) [reply decimal64math.go]: Still here

File: decimal64math.go (hunk 5 above)
Comment (juliaogris) [reply decimal64math.go]: Still here

File: decimal64math_test.go [hunk 7]
```diff
@@ -379,3 +379,93 @@ func BenchmarkDecimal64Sub(b *testing.B) {
...
+	b := MustParseDecimal64("0.000000000000001e384")
+	require.Equal(NegInfinity64, a.Sub(b))
+
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0.000000000000001e384")
+	require.Equal(Infinity64, a.Add(b))
+
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0.0000000000000001e384")
+	require.Equal(a, b.Add(a))
```
Comment (anzdaddy) [inline decimal64math_test.go]: I suggest a simpler layout. While I'm loath to write the code for you, this one's much easier to show than describe:
```go
	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
	require.Equal(Infinity64, a.Add(MustParseDecimal64("1e369")))
	require.Equal(a, a.Add(MustParseDecimal64("1e369")))
```
You should also test for more borderline edge cases, such as:

- 9.999999999999999e384 + 0.4999999999999999e369
- 9.999999999999999e384 + 0.5e369.

I'm not sure what the standard says here. After scale-matching, do you add then round or round then add, and how many digits do you hold during the calculation? We can use uint128T to make the calculations mathematically precise, but the standard might allow (or require) some precision to be thrown away early to support maximal efficiency and limitations of hardware implementations. Raise an issue to resolve this later.

Either way, the principle is to discern the exact boundary between stays-the-same and jumps-to-∞, and test the two values immediately above and below it.

File: decimal64math_test.go [hunk 8]
```diff
@@ -379,3 +379,93 @@ func BenchmarkDecimal64Sub(b *testing.B) {
...
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0")
+	require.Equal(a, b.Add(a))
+
+}
+
+func TestQuoOverflow(t *testing.T) {
+	require := require.New(t)
+
+	max := MustParseDecimal64("1e384") // largest dec64 number
```
Comment (anzdaddy) [inline decimal64math_test.go]: Not the largest. Ditto below.

Comment (anzdaddy) [reply decimal64math_test.go]: Actually, we should simply define `const Max64 = newFromParts(…)` in decimal64const.go. Also `Min64` (the smallest positive subnormal) and maybe even `MaxSubnormal64` and `MinNormal64`

File: decimal64math_test.go [hunk 9]
```diff
@@ -379,3 +379,93 @@ func BenchmarkDecimal64Sub(b *testing.B) {
...
+	b = MustParseDecimal64("0")
+	require.Equal(Zero64, a.Mul(b))
+
+	a = MustParseDecimal64("0") // largest dec64 number
+	b = MustParseDecimal64("100")
+	require.Equal(Zero64, a.Mul(b))
+
+}
+
+// func TestQuoUnderflow(t *testing.T) {
```
Comment (anzdaddy) [inline decimal64math_test.go]: If this is to be rehydrated in future, put a `// TODO: …` in front of it.

File: decimal64math.go [hunk 10]
```diff
@@ -133,12 +136,18 @@ func (d Decimal64) Quo(e Decimal64) Decimal64 {
...
+	if e == Zero64 || e == NegZero64 {
+		return infinities[sign1]
+	}
+
 	exp := exp1 - exp2 - 16
 	significand := umul64(10*decimal64Base, significand1).div64(significand2)
 	for significand.hi > 0 || significand.lo >= 10*decimal64Base {
 		exp++
 		significand = significand.divBy10()
+
```
Comment (anzdaddy) [inline decimal64math.go]: Unnecessary blank line.

File: decimal64.go [hunk 11]
```diff
@@ -133,7 +133,12 @@ func matchScales(exp1 int, significand1 uint64, exp2 int, significand2 uint64) (
 }
 
 func newFromParts(sign int, exp int, significand uint64) Decimal64 {
+	// Must be able to signify when significand is subnormal
 	s := uint64(sign) << 63
+
+	if significand >= decimal64Base && exp >= expMax {
+		return infinities[sign]
+	}
```
Comment (anzdaddy) [inline decimal64.go]: Do this test before initialising `s` to avoid doing unnecessary work.

File: decimal64math.go [hunk 12]
```diff
@@ -133,12 +136,18 @@ func (d Decimal64) Quo(e Decimal64) Decimal64 {
 		return zeroes[sign]
 	}
 
+	if e == Zero64 || e == NegZero64 {
```
Comment (anzdaddy) [inline decimal64math.go]: We need `(Decimal64).IsZero64() bool`. Raise an issue for later.

File: decimal64math_test.go [hunk 13]
```diff
@@ -379,3 +379,93 @@ func BenchmarkDecimal64Sub(b *testing.B) {
...
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0.000000000000001e384")
+	require.Equal(Infinity64, a.Add(b))
+
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0.0000000000000001e384")
+	require.Equal(a, b.Add(a))
+
+	a = MustParseDecimal64("9.999999999999999e384") // largest dec64 number
+	b = MustParseDecimal64("0")
```
Comment (anzdaddy) [inline decimal64math_test.go]: Use `Zero64`. And don't bother sticking it in a variable. Just use it directly in `require.Equal(…)`. Ditto below.

File: decimal64math_test.go [hunk 14]
```diff
@@ -379,3 +379,93 @@ func BenchmarkDecimal64Sub(b *testing.B) {
 		_ = x[i%len(x)].Sub(y[i%len(y)])
 	}
 }
+
+func TestAddOverflow(t *testing.T) {
```
Comment (anzdaddy) [inline decimal64math_test.go]: Let's also test underflow scenarios for Add/Quo, etc. (Min/2 = 0)

Comment (juliaogris) [reply decimal64math_test.go]: Added as [comment](https://github.com/anz-bank/decimal/issues/1#issuecomment-456246840) to #1 - let's get this PR merged.

File: decimal64math.go (hunk 12 above)
Comment (juliaogris) [reply decimal64math.go]: #15 - Done

File: decimal64.go [hunk 15]
```diff
@@ -133,7 +133,17 @@ func matchScales(exp1 int, significand1 uint64, exp2 int, significand2 uint64) (
...
 func newFromParts(sign int, exp int, significand uint64) Decimal64 {
+
+	for significand >= maxSig {
+		exp++
+		significand /= 10
+	}
+
+	if exp > expMax {
+		return infinities[sign]
+	}
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64.go]: newFromParts now simplified, input from arithmetic functions already normalised with renormalize and inf rounding/zeros returned before newFromParts is called

File: decimal64const.go [hunk 16]
```diff
@@ -36,8 +36,17 @@ var inf64 uint64 = 0x78 << 56
 // 1E15
 const decimal64Base = 1000 * 1000 * 1000 * 1000 * 1000
 
+// Max significand
+const maxSig = 9999999999999999
```
[the remark this answers is not in the document]
Comment (Joshcarp) [reply decimal64const.go]: maxSig = 10*decimal64Base-1

**Comment by anzdaddy** [comment]
We're overloading this PR. Let's tag HEAD and temporarily reset master back to ddcd461 so we can approve and merge it.

**Comment by Joshcarp** [comment]
oh yeah, 
couln't merge the pr, dont have write access
